File: web/tests/drift.test.ts
import { describe, expect, it } from 'vitest';
import { canOverride, driftRows } from '../src/lib/drift';
import { DeploymentDrift } from '../src/api/types';

const drift = (status: DeploymentDrift['report']['status'], overrideStatus?: DeploymentDrift['report']['status']): DeploymentDrift => ({
  report: {
    deploymentId: 'd1', versionId: 'v1', version: 1, strategyId: 's1', mode: 'PAPER', enabled: true, sizeMultiplier: 0.5, status,
    window: { maxTrades: 30, sessions: 60, from: '2026-08-01', to: '2026-09-10' },
    live: { trades: 30, winRate: 0.47, expectancyR: -0.07, profitFactor: 0.9, maxDrawdownR: 5.25 },
    backtest: { trades: 60, winRate: 0.61, expectancyR: 0.28, profitFactor: 1.8, maxDrawdownR: 4, split: 'OUT_OF_SAMPLE' },
    triggered: [], evidence: [], assessedAt: '2026-09-10T10:00:00Z',
  },
  state: { deploymentId: 'd1', status, actedStatus: status, triggered: [], overrideStatus, updatedAt: '2026-09-10T10:00:00Z' },
  history: [],
});

describe('drift helpers', () => {
  it('lays out the PRD 25 comparison', () => {
    const rows = driftRows(drift('DEGRADING'));
    expect(rows.find((r) => r.metric === 'Win rate')).toEqual({ metric: 'Win rate', backtest: '61%', live: '47%' });
    expect(rows.find((r) => r.metric === 'Max drawdown')).toEqual({ metric: 'Max drawdown', backtest: '4.0R', live: '5.3R' });
    expect(rows.find((r) => r.metric === 'Expectancy')!.live).toContain('0.07');
    const noBacktest = { ...drift('INSUFFICIENT_DATA'), report: { ...drift('INSUFFICIENT_DATA').report, backtest: undefined } };
    expect(driftRows(noBacktest).every((r) => r.backtest === '—')).toBe(true);
  });
  it('offers an override for WATCH or worse once', () => {
    expect(canOverride(drift('DEGRADING'))).toBe(true);
    expect(canOverride(drift('DEGRADING', 'DEGRADING'))).toBe(false);
    expect(canOverride(drift('FAILED', 'DEGRADING'))).toBe(true);
    expect(canOverride(drift('HEALTHY'))).toBe(false);
    expect(canOverride({ ...drift('WATCH'), state: undefined })).toBe(false);
  });
});

File: tui/internal/ui/drift_test.go
package ui

import (
	"strings"
	"testing"

	"hejje.money/tui/internal/api"
)

func TestRenderDriftShowsStatusSidesAndCriteria(t *testing.T) {
	pf := 1.8
	d := api.StrategyDrift{Enabled: true, Deployments: []api.DeploymentDrift{{
		Report: api.DriftReport{DeploymentID: "dep1", Version: 2, Mode: "PAPER", Enabled: true, SizeMultiplier: 0.5, Status: "DEGRADING",
			Live:      api.DriftSide{Trades: 30, WinRate: 0.47, ExpectancyR: -0.07, MaxDrawdownR: 5.2},
			Backtest:  &api.DriftSide{Trades: 60, WinRate: 0.61, ExpectancyR: 0.28, ProfitFactor: &pf, MaxDrawdownR: 4, Split: "OUT_OF_SAMPLE"},
			Triggered: []string{"win rate 47% vs 61% backtest is unlikely by chance (p = 0.030 < 0.05)"}},
		State: &api.DriftState{Status: "DEGRADING", OverrideStatus: "DEGRADING", OverrideBy: "admin", OverrideReason: "reviewed"},
	}}}
	d.Deployments[0].Report.Window.MaxTrades = 30
	out := RenderDrift(d)
	for _, want := range []string{"dep1 v2 PAPER enabled  DEGRADING  size x0.50", "BACKTEST (OOS)", "TRAILING 30", "61%", "47%", "+0.28R", "-0.07R", "1.80",
		"5.2R", "! win rate 47%", "override: DEGRADING by admin — reviewed"} {
		if !strings.Contains(out, want) {
			t.Fatalf("missing %q in\n%s", want, out)
		}
	}
	if !strings.Contains(RenderDrift(api.StrategyDrift{}), "drift monitor disabled") {
		t.Fatal("disabled render")
	}
}
